README.md
# 吉珠小助手

[![GitHub](https://img.shields.io/badge/license-WTFPL-blue.svg?style=flat-square)](https://zh.wikipedia.org/wiki/WTFPL)


> 一个多年的心愿

## 预览
微信搜索同名小程序或者扫描下方小程序码即可预览

![小程序二维码](/images/qrcode.jpg)

## 开发
> 吉珠小助手的后台 mock 地址为：  
> https://easy-mock.com/mock/5c149dff1ee30f317685ba2d  
> 为了获得较为完善的开发体验，可以在 config.ts 中填入该地址

开发环境为 Node.js v10.13.0 和 NPM v6.4.1

开发前请确保安装了 Node.js 和 NPM

1. 安装 [Taro](https://taro.js.org/)
2. Clone 本项目 `git clone https://github.com/xyxc0673/jza-taro.git`
3. 重命名 `project.config.sample.json` 为 `project.config.json`
4. 重命名 `src/utils/config.sample.ts` 为 `src/utils/config.ts`
5. 运行 `npm install`
6. 运行 `npm run dev:weapp`
7. 使用 [微信开发者工具](https://developers.weixin.qq.com/miniprogram/dev/devtools/download.html) 打开本项目
8. 使用顺手的开发工具，享受编码的乐趣 ( :

## 功能
> 一则令人遗憾的消息：  
> 由于没有内网服务器，宿舍电费功能变得遥遥无期。  

- [x] 教务系统
  - [x] 课表
    - [x] 手动添加
    - [x] 班级课表
  - [x] 成绩
    - [x] 成绩分析
- [x] 校园卡
  - [x] 余额
  - [x] 消费记录
- [x] 图书馆
  - [x] 书目检索
  - [x] 个人中心
    - [x] 借阅查询
    - [x] 续借书籍
- [ ] 宿舍电费 ( ? )
  - [ ] 余量查询
  - [ ] 购电记录
  - [ ] 用电记录
  - [ ] 充值记录
- [x] 通知
- [ ] 校历
- [ ] 百科

## 感谢

- [Taro](https://taro.js.org/)
- [Iconfont](http://www.iconfont.cn/)
- [We川大](https://github.com/mohuishou/scuplus-wechat)
- [微吉风](https://github.com/choyri/WeGifun)
- [课表](http://kb.mayuko.cn/)

## 许可
[WTFPL](https://zh.wikipedia.org/wiki/WTFPL)

CHANGELOG.md
# 更新日志
## v0.4.4(2019-09-09)
- **新增** 设置关于页面增加交流QQ群
- **调整** 班级课表可查询2019级以及默认选中最新年级

## v0.4.3(2019-05-07)
- **优化** 细节调整

## v0.4.2(2019-03-22)
- **新增** 支持继续教育学院
- **新增** 课表周六、周日显示开关
- **修复** 八九大节上课时间显示不正确
- **修复** 班级课表部分学院学期信息错误

## v0.4.1(2019-03-16)
- **新增** 课表设置背景时表头日期栏透明开关
- **修复** 特殊教室上课时间错误
- **修复** 少数情况下课程格子显示有问题

## v0.4.0(2019-03-12)
- **新增** 我的课表自定义背景图片功能
- **新增** 首页今日课表可开启显示上课时间
- **优化** 加快书目检索加载信息的速度
- **优化** 今日课表布局与增加显示教师设置
- **优化** 教务系统班级课表的获取界面与逻辑

## v0.3.8(2019-03-01)
- **新增** 班级课表自动选择当前学年学期
- **新增** 设置页面小程序分享按钮
- **修复** 当前学期判断错误
- **优化** 部分界面布局与显示效果

## v0.3.7(2019-02-19)
- **新增** 首页部分卡片点击可直达相应功能
- **修复** 周数可能导致无法获取和正常显示课表
- **修复** 周数可能导致无法获取和正常显示课表
- **优化** 更新检测程序
- **优化** 部分文案和显示效果

## v0.3.6(2019-01-14)
- **修复** 无法添加自定义课程
- **优化** 班级课表的存储结构
- **优化** 开学日期的获取方式
- **优化** 旧版本的升级检测方式

## v0.3.5 (2019-01-08)
- **新增** 新版本特性提示
- **新增** 旧版本判断与升级
- **新增** 班级课表支持保存整个课表
- **修复** 我的课程表滑动切换周数时周数越界

## v0.3.4 (2019-01-03)
- **修复** 设置页面不能正常解绑账号
- **修复** 我的课程表课程格子宽度不统一
- **优化** 部分样式和逻辑代码
- **优化** 显示下周课程改为显示非本周课程

## v0.3.3 (2018-12-23)
- **优化** 我的课程表布局代码
- **优化** 图书检索书籍封面获取方式

## v0.3.2 (2018-12-16)
- **修正** 拼写错误
- **修复** 我的课程表表头日期不正确
- **优化** 重构部分图书馆个人中心代码
- **优化** 图书馆个人中心借阅历史开启分页显示
- **优化** 班级课表未选择时的提示从专业改为学院

## v0.3.1 (2018-12-14)
- **修复** 显示的周数与正确的周数相差一周的问题
- **修复** 班级课表中的课程选课人数显示 undefined
- **修复** Android 设备和 iPhone 设备课表样式显示不一致
- **修复** 与教务系统相关的页面可能无法正常地选择学年或学期
- **优化** 检索完课表进入我的课表时不再关闭当前页面

## v0.3.0 (2018-12-12)
- **新增** 班级课表查询及其课程添加
- **修复** 课程表表头高亮显示不正确
- **修复** 课程表格子可能会错位的问题
- **修复** 两个逻辑与写成一个逗号的错误
- **修复** 图书馆个人中心操作后信息未自动更新
- **优化** 前后端变量写法统一为驼峰式
- **优化** 完善登录引导机制

## v0.2.0 (2018-12-05)
- **新增** 图书馆个人中心
- **新增** 左右滑动我的课程表主体可切换上下周课程
- **修复** 一些学院拥有不同的学年制度所产生的问题
- **修复** 图书查询第一次加载更多时出现重复的记录
- **修复** 我的课程表表头日期中月份不正确的问题
- **优化** 部分样式、文案与代码结构

## v0.1.1 (2018-11-29)
- **新增** 显示下周课程及其选项开关
- **修复** 课表设置页面检索课表后今日课表不自动刷新

## v0.1.0 (2018-11-26)
- 小程序初版完成

src/services/schedule.ts
import Taro from '@tarojs/taro'

import account from './account'

import utils from '../utils/utils'
import request from '../utils/request'

interface ISchedule {
  index: Array<number>
  courseName: string,
  flex: number,
}

class Schedule {
  static getSchoolStartDate () {
    return Taro.getStorageSync('schoolStartDate')
  }
  
  static async getCurrWeek (refresh = false)  {

    let schoolStartDate: string = this.getSchoolStartDate()

    if (refresh || !schoolStartDate) {
      const res = await request.jwSchoolStartDate()
      if (!res) { return }
      schoolStartDate = res.data.data.date
    }

    Taro.setStorageSync('schoolStartDate', schoolStartDate)

    return Math.ceil((new Date().getTime() - new Date(schoolStartDate).getTime()) / (1000 * 3600 * 24 * 7))
  }
  
  static getDayDate (week) {
    const schoolStartDate: Date = new Date(this.getSchoolStartDate())
    const days: Array<any> = []
  
    schoolStartDate.setDate(schoolStartDate.getDate() + (week - 1) * 7 - 1)
  
    for (let i = 0; i < 7; i ++) {
      const timestamp = schoolStartDate.setDate(schoolStartDate.getDate() + 1)
      const datetime = new Date(timestamp)
      days.push({date: (datetime.getMonth() + 1) + '-' + datetime.getDate(), day: '周' + utils.replaceToChinese(datetime.getDay()), dayInt: datetime.getDay()})
    }
  
    return days
  }

  static colors = [
    'rgb(196, 98, 67)', // 照柿
    'rgb(208, 16, 76)', // 韓紅花
    'rgb(0, 137, 108)', // 青竹
    'rgb(180, 165, 130)', // 利休白茶
    'rgb(58, 143, 183)', // 千草
    'rgb(138, 107, 190)', // 藤紫
    'rgb(148, 122, 109)', // 胡桃
    'rgb(147, 150, 80)', // 柳茶
    'rgb(144, 180, 75)', // 鶸萌黄
    'rgb(51, 103, 116)', // 錆御納戸
    'rgb(38, 135, 133)', // 青碧
    'rgb(34, 125, 81)', // 緑
  ]

  static timeTable = [
    ['08:20', '09:05'],
    ['09:15', '10:00'],

    [
      ['10:20', '10:30'],
      ['11:05', '11:15'],
    ],

    [
      ['11:15', '11:25'],
      ['12:00', '12:10'],
    ],

    ['14:00', '14:45'],
    ['14:55', '15:40'],

    ['16:00', '16:45'],
    ['16:55', '17:40'],

    ['18:40', '19:25'],
    ['19:35', '20:20'],

    ['20:30', '21:15'],
    ['21:25', '22:10']
  ]

  static notCurrWeekCourseColor = 'rgba(120,125,123, 0.5)' // 素鼠

  static async Get (year: number = 0, semester: number = 0) {
    if (!year || !semester) {
      const ret = account.calYearSemester()
      year = ret['year']
      semester = ret['semester']
    }

    const response = await request.jwSchedule({year: year, semester: semester})
    if (!response.data || response.data.data.code == -1) {
      return false
    }
    
    Taro.setStorageSync('schedule', response.data.data.schedule)
    Taro.showToast({title: '获取课表成功', icon: "none"})
    return true
  }

  static GetFormStorage () {
    const schedule = Taro.getStorageSync('schedule')
    if (!schedule) {
      return false
    }
    return schedule
  }

  static initFrame () {
    /*
    生成一个 8 * 12 的矩阵
    矩阵第一行是 [1, 2, 3, 4, 5, 6, 7, 8,  9, 10, 11, 12]
    其余行均是 ISchedule * 12 的数组
    */
    let schedule: Array<any> = [[1, 2, 3, 4, 5, 6, 7, 8,  9, 10, 11, 12]]
    for (let i = 0; i < 7; i++) {
      let courseBlock: Array<ISchedule> = []
      for (let j = 0; j < 12; j++) {
        courseBlock[j] = {
          index: [i+1, j+1],
          courseName: '',
          flex: 1,
        }
      }
      schedule.push(courseBlock)
    }
    return schedule
  }

  static InitSchedule (rawSchedule: Array<any>, week: number = -1, day: number = -1, displaNotCurrentWeekCourse: boolean = false) {
    const sameScheduleSameColor = {}
    const colorLength = this.colors.length
    let schedule: Array<any> = []

    const setting = Taro.getStorageSync('setting')
    displaNotCurrentWeekCourse = displaNotCurrentWeekCourse || setting.displaNotCurrentWeekCourse

    if (day === -1) {
      schedule = this.initFrame()
    }
    
    let colorIndex = 0

    // 初始化课程表
    for (let s of rawSchedule) {
      let currWeekCourse = false

      if (!s.during) {
        continue
      }

      const allWeek = s.during.split(',')
      s.firstWeek = parseInt(allWeek[0])

      // 判断是否为本周课程或下周课程
      for (let w of allWeek) {
        let wInt = parseInt(w)
        if (wInt == week) {
          currWeekCourse = true
          break
        } else if (wInt > week) {
          break
        }
      }

      s.sessionArray = s.session.split(',')
      s.currWeekFlag = currWeekCourse

      s.timeTable = this.getTimeTable(s).join('-')
      s.sessionText = s.sessionArray[0] + '-' + s.sessionArray[s.sessionArray.length-1]

      // 调用来自于首页的今日课表
      if (day !== -1) {
        // 如果是本周课程且是当日课程
        if (parseInt(s.day) == day &&  currWeekCourse) {
          schedule.push(s)
        }
        // 无论如何，跳过本次循环
        continue
      }

      // 显示非本周课程设置未开启, 跳过非本周课程
      if (!displaNotCurrentWeekCourse && !currWeekCourse) {
        continue
      }
      
      const firstSession = s.sessionArray[0] - 1
      const classInCurrCell = schedule[parseInt(s.day)][firstSession]

      // 计算课程首周与本周的周差
      const sDistance = Math.abs(s.firstWeek - week)
      const cDistance = Math.abs(classInCurrCell.firstWeek - week)
      
      // 课程矩阵中将要放置的格子已经有课程，以下两个情况跳过渲染
      // 1. 格子中的课程是本周要上的课
      // 2. 格子中的课程首周比当前课程的要晚
      if (classInCurrCell.flex > 1 && (classInCurrCell.currWeekFlag || sDistance > cDistance)) {
        continue
      }

      let x = parseInt(s.day)
      let y = firstSession

      // 判断当前cell是否已经被划分为不显示的格子，即已经被某个课程撑开而隐藏，但是有可能该课程撑开的大小会占用到当前循环课程的位置
      if (classInCurrCell.flex === 0) {
        // 跳过虽然格子被占用，但是当前课程不是本周课程
        if (!s.currWeekFlag) {
          continue
        }

        // 向上寻找占用格子的课程
        while ((--y) >= 0) {
          if (schedule[x][y].flex > 1) {
            // 缩小课程格子
            schedule[x][y].flex -= s.sessionArray.length
          }
        }
      }

      s.duringText = allWeek[0] + '-' + allWeek[allWeek.length - 1] // start week to end week

      // 课程块随机上色，并为同一课程上相同色
      if (currWeekCourse) {
        s.color = sameScheduleSameColor[s.courseName] || this.colors[(colorIndex++) % colorLength]
        sameScheduleSameColor[s.courseName] = s.color
      } else {
        s.color = this.notCurrWeekCourseColor
      }
      
      schedule = this.doFlex(schedule, s)
    }

    return schedule
  }

  static getTimeTable (s) {
    const sessionArray = s.sessionArray
    const startSession = sessionArray[0]
    const endSession = sessionArray[sessionArray.length-1]
    const isSpecial = s.location.match(/二教|实验|室/g) === null ? 0 : 1

    const getTime = (session, isStart) => {
      const sessionInt = parseInt(session)
      const tmp = this.timeTable[sessionInt - 1]

      if (sessionInt === 3 || sessionInt === 4) {
        return tmp[isStart ? 0 : 1][isSpecial]
      }

      return tmp[isStart ? 0 : 1]
    } 

    const startTime = getTime(startSession, true)
    const endTime = getTime(endSession, false)

    return [startTime, endTime]
  }

  static doFlex (schedule, s) {
    const dayInt = parseInt(s.day)

    // 将与该课程上课节数有关的 cell 的 flex 置 0
    s.sessionArray.forEach((i) => {
      schedule[dayInt % 8][i - 1].flex = 0;
    })

    // 用于课程第一个 cell 的 flex 值，予以撑开所占节数
    s.flex = s.sessionArray.length
    // 将与该课程上课节数有关的第一个 cell 替换为 该课程
    schedule[dayInt % 8][s.sessionArray[0] - 1] = s

    return schedule
  }
}

export default Schedule